driver-core/src/test/unit/com/mongodb/internal/connection/AsynchronousClusterEventListenerTest.java

```diff
@@ -0,0 +1,171 @@
+        listener.clusterOpening(clusterOpeningEvent);
+        assertEquals(clusterOpeningEvent, targetListener.take());
+
+        ClusterDescriptionChangedEvent clusterDescriptionChangedEvent = new ClusterDescriptionChangedEvent(clusterId,
+                new ClusterDescription(ClusterConnectionMode.SINGLE, ClusterType.STANDALONE, Collections.emptyList()),
+                new ClusterDescription(ClusterConnectionMode.SINGLE, ClusterType.STANDALONE, Collections.emptyList()));
+        listener.clusterDescriptionChanged(clusterDescriptionChangedEvent);
+        assertEquals(clusterDescriptionChangedEvent, targetListener.take());
+
+        ServerHeartbeatStartedEvent serverHeartbeatStartedEvent = new ServerHeartbeatStartedEvent(connectionId);
+        listener.serverHearbeatStarted(serverHeartbeatStartedEvent);
+        assertEquals(serverHeartbeatStartedEvent, targetListener.take());
+
+        ServerHeartbeatSucceededEvent serverHeartbeatSucceededEvent = new ServerHeartbeatSucceededEvent(connectionId, new BsonDocument(),
+                1, true);
+        listener.serverHeartbeatSucceeded(serverHeartbeatSucceededEvent);
+        assertEquals(serverHeartbeatSucceededEvent, targetListener.take());
+
+        ServerHeartbeatFailedEvent serverHeartbeatFailedEvent = new ServerHeartbeatFailedEvent(connectionId, 1, true, new IOException());
+        listener.serverHeartbeatFailed(serverHeartbeatFailedEvent);
+        assertEquals(serverHeartbeatFailedEvent, targetListener.take());
+
+        ServerOpeningEvent serverOpeningEvent = new ServerOpeningEvent(serverId);
+        listener.serverOpening(serverOpeningEvent);
+        assertEquals(serverOpeningEvent, targetListener.take());
+
+        ServerDescriptionChangedEvent serverDescriptionChangedEvent = new ServerDescriptionChangedEvent(serverId,
+                ServerDescription.builder().address(new ServerAddress()).type(UNKNOWN).state(CONNECTING).build(),
+                ServerDescription.builder().address(new ServerAddress()).type(STANDALONE).state(CONNECTED).build());
+        listener.serverDescriptionChanged(serverDescriptionChangedEvent);
+        assertEquals(serverDescriptionChangedEvent, targetListener.take());
+
+        ServerClosedEvent serverClosedEvent = new ServerClosedEvent(serverId);
+        listener.serverClosed(serverClosedEvent);
+        assertEquals(serverClosedEvent, targetListener.take());
+
+        ClusterClosedEvent clusterClosedEvent = new ClusterClosedEvent(clusterId);
+        listener.clusterClosed(clusterClosedEvent);
+        assertEquals(clusterClosedEvent, targetListener.take());
+
+        // The thread should die after publishing the ClusterClosedEvent
+        listener.getPublishingThread().join(5000);
+    }
+
+    private static final class AllClusterEventListener implements ClusterListener, ServerListener, ServerMonitorListener {
+        private final BlockingQueue<Object> lastEvent = new ArrayBlockingQueue<>(1);
```

Comment:
If we use `SynchronousQueue` here instead of `ArrayBlockingQueue` (it's also a `BlockingQueue`), then `testEventsPublished` will also verify that `AsynchronousClusterEventListener` delivers events asynchronously: if it were to deliver synchronously, then the test thread would have hanged on sending events via `AsynchronousClusterEventListener`.